src/sqlancer/clickhouse/oracle/ClickHouseTLPHavingOracle.java

```diff
@@ -0,0 +1,78 @@
+package sqlancer.clickhouse.oracle;
+
+import sqlancer.DatabaseProvider;
+import sqlancer.IgnoreMeException;
+import sqlancer.Randomly;
+import sqlancer.TestOracle;
+import sqlancer.clickhouse.ClickHouseErrors;
+import sqlancer.clickhouse.ClickHouseProvider;
+import sqlancer.clickhouse.ClickHouseSchema;
+import sqlancer.clickhouse.ClickHouseVisitor;
+import sqlancer.clickhouse.ast.*;
+import sqlancer.clickhouse.gen.ClickHouseCommon;
+import sqlancer.clickhouse.gen.ClickHouseExpressionGenerator;
+
+import java.sql.SQLException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ClickHouseTLPHavingOracle implements TestOracle {
```

Comment:
It is not necessary for this PR, but in the future you might want to consider inheriting from `ClickHouseTLPBase` and use the existing functionality there. I already implemented this for some of the other DBMS.